packages/jsapi-nodejs/src/NodeHttp2gRPCTransport.ts

```diff
@@ -0,0 +1,194 @@
+   * @param headers Request headers
+   * @returns The created http2 client stream
+   */
+  createRequest = (
+    headers: Record<string, string> | null
+  ): http2.ClientHttp2Stream => {
+    const url = new URL(this.options.url);
+
+    logger.debug('createRequest', url.pathname);
+
+    const req = this.session.request({
+      ...headers,
+      ':method': 'POST',
+      ':path': url.pathname,
+    });
+
+    req.on('response', (responseHeaders, _flags) => {
+      const headersRecord: Record<string, string | string[]> = {};
+
+      // strip any undefined headers or keys that start with `:`
+      Object.keys(responseHeaders).forEach(name => {
+        if (responseHeaders[name] != null && !name.startsWith(':')) {
+          headersRecord[name] = responseHeaders[name];
+        }
+      });
+
+      this.options.onHeaders(headersRecord, Number(responseHeaders[':status']));
+    });
+
+    // Note that `chunk` is technically a `Buffer`, but the `Buffer` type defined
+    // in @types/pouchdb-core is outdated and incompatible with latest `Uint8Array`
+    // types. Since `Buffer` inherits from `Uint8Array`, we can get around this
+    // by just declaring it as a `Uint8Array`.
+    req.on('data', (chunk: Uint8Array) => {
+      this.options.onChunk(chunk);
+    });
+    req.on('end', () => {
+      this.options.onEnd();
+    });
+    req.on('error', err => {
+      this.options.onEnd(err);
+    });
+
+    return req;
+  };
+
+  /**
+   * Starts the stream, sending metadata to the server.
+   * @param metadata - the headers to send the server when opening the connection
+   */
+  start(metadata: { [key: string]: string | Array<string> }): void {
+    logger.debug('start', metadata.headersMap);
+
+    if (this.request != null) {
+      throw new Error('start called more than once');
+    }
+
+    const headers: Record<string, string> = {};
+    Object.entries(metadata).forEach(([key, value]) => {
+      headers[key] = typeof value === 'string' ? value : value.join(', ');
+    });
+
+    this.request = this.createRequest(headers);
+  }
+
+  /**
+   * Sends a message to the server.
+   * @param msgBytes - bytes to send to the server
+   */
+  sendMessage(msgBytes: Uint8Array): void {
+    logger.debug('sendMessage', msgBytes);
+    assertNotNull(this.request, 'request is required');
+
+    this.request.write(msgBytes);
+  }
+
+  /**
+   * "Half close" the stream, signaling to the server that no more messages will
+   * be sent, but that the client is still open to receiving messages.
+   */
+  finishSend(): void {
+    logger.debug('finishSend');
+    assertNotNull(this.request, 'request is required');
+    this.request.end();
+  }
+
+  /**
+   * End the stream, both notifying the server that no more messages will be
+   * sent nor received, and preventing the client from receiving any more events.
+   */
+  cancel(): void {
+    logger.debug('cancel');
+    assertNotNull(this.request, 'request is required');
+    this.request.close();
```

Comment:
Should we reset `request` to `null` here? Probably not needed, if `request.write` handles closed requests fine.

Reply:
Good question. @niloc132 do you have any thoughts on this one?

Reply:
It looks safe to me to not null out - but nulling it out does come with the benefit of clearly showing why an error happened if it is reused somehow. I'm certainly not against that style of defensive coding, but I think we'd want to update a few other places too to reflect this.

GC-wise, I don't see any issue with the current implementation.

Reply:
Opting to leave as-is for now to avoid adding complexity. Setting to null will require changes to `start` which assumes this stays set to know if it has already been called.